File: daybreak/src/daybreak/core/theme_registry.py
from daybreak.themes import THEME_LIBRARY, get_theme_palette
from .theme_model import normalize_mode, validate_tokens
from .theme_transform import palette_to_tokens


class ThemeRegistry:
    def list_themes(self):
        return sorted(THEME_LIBRARY.keys())

    def get_palette(self, theme_name: str, mode: str) -> dict:
        normalized_mode = normalize_mode(mode)
        palette_set = get_theme_palette(theme_name)
        if normalized_mode not in palette_set:
            raise ValueError(f"Theme '{theme_name}' does not provide mode '{normalized_mode}'.")
        return palette_set[normalized_mode]

    def get_tokens(self, theme_name: str, mode: str) -> dict:
        palette = self.get_palette(theme_name, mode)
        tokens = palette_to_tokens(palette, mode)
        validate_tokens(tokens)
        return tokens

    def get_theme(self, theme_name: str, mode: str) -> dict:
        normalized_mode = normalize_mode(mode)
        tokens = self.get_tokens(theme_name, normalized_mode)
        palette = self.get_palette(theme_name, normalized_mode)
        return {
            "id": theme_name.lower().replace(" ", "-"),
            "name": theme_name,
            "mode": normalized_mode,
            "tokens": tokens,
            "palette": palette,
        }

File: src/core/themeRegistry.ts
import { Palette } from '../colors';
import { THEME_LIBRARY, getThemePalette } from '../themes';
import { normalizeMode, validateTokens } from './themeModel';
import { paletteToTokens, paletteToAccentTokens } from './themeTransform';

export class ThemeRegistry {
  listThemes(): string[] {
    return Object.keys(THEME_LIBRARY).sort();
  }

  getPalette(themeName: string, mode: string): Palette {
    const normalizedMode = normalizeMode(mode);
    const paletteSet = getThemePalette(themeName);
    const palette = paletteSet[normalizedMode as 'light' | 'dark'];
    if (!palette) {
      throw new Error(`Theme '${themeName}' does not provide mode '${normalizedMode}'.`);
    }
    return palette;
  }

  getTokens(themeName: string, mode: string): Record<string, string> {
    const palette = this.getPalette(themeName, mode);
    const tokens = paletteToTokens(palette, mode);
    validateTokens(tokens);
    return tokens;
  }

  getAccentTokens(themeName: string, mode: string): Record<string, string> {
    const palette = this.getPalette(themeName, mode);
    return paletteToAccentTokens(palette, mode);
  }

  getTheme(themeName: string, mode: string): object {
    const normalizedMode = normalizeMode(mode);
    const tokens = this.getTokens(themeName, normalizedMode);
    const accentTokens = this.getAccentTokens(themeName, normalizedMode);
    const palette = this.getPalette(themeName, normalizedMode);
    return {
      id: themeName.toLowerCase().replace(/ /g, '-'),
      name: themeName,
      mode: normalizedMode,
      tokens,
      accent_tokens: accentTokens,
      palette,
    };
  }
}
